### PR TITLE
CAMEL-9570:  Blueprint service proxies aren't used

### PR DESCRIPTION
**Pull request by hqstevenson**
This PR has two main pieces.  First, the CamelDependenciesFinder was removed from the CamelNamespaceHandler.  This code caused the issue described in CAMEL-9570 as well as CAMEL-10394.  The net effect of removing this class is the CamelContext may attempt to start when a service isn't available because the reference hasn't been specified in the XML.  This is better than what happens today which is service references sometimes get registered for services that don't exist - therefore, starting Blueprint Context times-out waiting for service references.  This one actually bit me 18-mo ago with a customer - I'm just didn't know at the time what was causing it.

The second piece of the PR is changing the BlueprintCamelContext so it starts after the BlueprintContainer is created (on the BlueprintEvent.CREATED).  This fixes some startup issues.  Basically what was happening is if the serviceChanged method (which previously was used to start the camel context) threw a RuntimeException, the Karaf container would spin until a stack overflow occurred.  I didn't see this behavior when the context is started after the BlueprintContainer is fully initialized.

The only thing I changed in the CamelContextFactoryBean was removing a space that caused a Checkstyle error.

### CODE REVIEW

**Comment by davsclaus** [comment]
I dont think this is a good idea to remove the CamelDependenciesFinder which has been in use for many many users in camel-blueprint. This is scanning your routes and detecting which Camel components / data formats etc are in use and has a dependency on them, so Camel wont blow up on startup.

I suggest to break this PR into two, one that does that other change about starting blueprint in other events, which seems maybe as something that can be doable.

**Comment by hqstevenson** [comment]
Thanks for looking at this Claus - I can do that. I did the PR in several commits so it could be cherry-picked, but I can create another PR.

On the CamelDependenciesFinder - I haven’t done a huge amount of testing with it, but what I’ve seen is that it isn’t helping.  What it does is register a service reference as a dependence for the BlueprintCamelContext.  What I’ve seen happen is the BlueprintCamelContext fails to start, waiting on some dependency - I’ve never seen it fail because it couldn’t find a service.

I would like to know about any test cases out there that depend on this functionality - I couldn’t find any.  I’d really like to have some good tests to validate I’m not breaking anything.

I have another version that still calls the CamelDependenciesFinder, but it makes a temporary copy of the context to avoid messing up the Blueprint configuration - would that be more acceptable?
\

**Comment by davsclaus** [comment]
Create a Camel blueprint XML file that uses a Camel component that doesnt come from camel-core and then deploy the xml file. The dependency finder will detect this new component and make a service reference, so osgi will graceful wait for the service to be ready.

You can then see its in GraceWait mode (or what its called) if you do a karaf list.

Then you can install that component using features:install and then the blueprint XML file is then started and run the Camel application.

**Comment by hqstevenson** [comment]
I removed the commit that removed the CamelDependenciesFinder.  I'm working on another PR that will address that, so I think this PR is good to merge.

File: components/camel-blueprint/src/main/java/org/apache/camel/blueprint/BlueprintCamelContext.java [hunk 1]
```diff
@@ -143,23 +143,80 @@ public String getComponentDocumentation(String componentName) throws IOException
...
+            }
+
+            LOG.debug(String.format("blueprintEvent{ replay=%b type=%s bundle=%s } : ", event.isReplay(), eventTypeString, event.getBundle().getSymbolicName()));
+        }
+
+
+        if (!event.isReplay() && this.getBundleContext().getBundle().getBundleId() == event.getBundle().getBundleId()) {
+            if (event.getType() == BlueprintEvent.CREATED) {
+                try {
+                    this.start();
```
Comment (davsclaus) [inline components/camel-blueprint/src/main/java/org/apache/camel/blueprint/BlueprintCamelContext.java]: The old code was calling maybeStart that does some checks which I would like to keep using.

**Comment by hqstevenson** [comment]
Good catch!!  I amended the last commit and replaced the start call with maybeStart- which was a mistake because that only fixed a check style issue.  Hopefully amending the wrong commit isn't an issue.

**Comment by davsclaus** [comment]
This causes problems such as `camel-example-mybatis` will fail now.

I am not keen on this change as it breaks behavior for end users who rely on current behavior. We have no real problems reported by others and blueprint has been working pretty well. OSGi is hard.

This startup code/behavior has been in use for many years and thus "it works" for users.

**Comment by hqstevenson** [comment]
Wow - I guess I'll have to look at that example and see what is going on.  The change in the startup behavior shouldn't really be visible to any components running in the context, so that really surprises me.

I don't want to break current users either, so I guess we can mark CAMEL-9570 as "won't fix".  I'll continue to address the problems with camel-blueprint the way I have in the past - with a custom CamelContext.  This isn't really a big deal - I just don't like maintaining code in a customer codebase that I think should be part of the product.

How would you feel about adding another class - something like BlueprintCamelContextBean - that would behave the way I'd expect with a Java RouteBuilder and such.  You wouldn't use the Camel namespace, so it shouldn't cause any problems that way, but it would give customers the option of using Blueprint to manage dynamic services for them.  That would also be a good place to add the other enhancement my customers want - having the context suspend/resume when a required service disappears/reappears.

If that sounds more reasonable/safer to you as well, I'll open another JIRA for that enhancement.  Any suggestions on the name of the "bean" would be greatly appreciated as well.

**Comment by davsclaus** [comment]
So I debugged that example a bit and the difference is that <bean> that depends on <camelContext> now don't receive a started CamelContext injected. Without this change the CamelContext was started beforehand the bean is in use. The example uses a bean with an init/destroy method to setup the bean using CamelContext.

**Comment by davsclaus** [comment]
@hqstevenson I am not sure the community understand your use-case for that customer.

In general we should be a bit wary to change something based on a single need from a single user. I am frankly not to happy to add stuff to camel-blueprint for one special use-case.

All the blueprint / servicemix / talend / fuse / and many other users of Camel with blueprint have been working well so far. Yes over the years we had some issues and problems that has been sorted.

And I am afraid some of the hard core OSGi users are not so active anymore as they used to be like gnodet, jb, and whatelse.

Maybe what goes wrong is that this PR has a title that talks about service proxies and then there is 2 different commits. 

There is 7 JIRA tickets about camel-blueprint
https://issues.apache.org/jira/issues/?jql=component%20%3D%20camel-blueprint%20AND%20project%20%3D%20CAMEL%20AND%20resolution%20%3D%20Unresolved%20ORDER%20BY%20updated%20DESC

**Comment by hqstevenson** [comment]
I agree - I confused matters by adding the change to the startup/shutdown behavior in this JIRA.  I apologize - I shouldn't have mixed things up like that.  I should've opened another JIRA to cover the startup behavior - I'll do that.  It may never get implemented, but at least the reasoning why will be documented in the JIRA - hopefully saving others from going down the path I've been down.

The core issue I'm still fighting is that the Blueprint proxies are not used as long as the CamelDependenciesFinder is forcing the premature creation of objects.  I'll have to work some more on this to see if I can make it work without completely removing the CamelDependenciesFinder.

I also agree that the core shouldn't change based on the needs of a single customer (unless it's a true bug of course).  The problem I'm facing is I've hit this same situation with three customers now - but it all has to do with using Java RouteBuilders with the BlueprintCamelContext - I haven't seen any issues when using XML to define the routes.  That's why I was suggesting that we add another "BlueprintRouteBuilder" that can be configured without using the Camel Blueprint namespace - that would solve my customers issues, and maintain backward compatibility.

The startup method just seemed a little unorthodox for OSGi - I've been doing OSGi for quite some time, and I'm used to components that run under the control of Blueprint to start when the Blueprint context is fully created and ready - not before.  This issue this causes is when for some reason the context fails to start and throws an exception, it triggered an infinite loop which finally resulted in a stack overflow.  This made it very difficult to debug since the logs had rolled several times, and finding the actual problem was a bit of a task.

**Comment by hqstevenson** [comment]
I created a JIRA for the startup change improvement, opened a PR for it - I'm closing this PR and I'll work on another one that fixes the service issue in a backward-compatible fashion.

**Pull request closed without merging**